h2database#308

x7FFFFFFF opened the pull request:
https://github.com/h2database/h2database/issues/308
Mode MySQL and LAST_INSERT_ID with argument #308

paramonov-alex committed:
h2database#308  https://github.com/h2database/h2database/issues/308   Mode MySQL and LAST_INSERT_ID with argument #308

katzyn commented on h2/src/main/org/h2/expression/function/Function.java:
```diff
@@ -390,7 +390,7 @@
                 0, Value.LONG);
         addFunctionNotDeterministic("LAST_INSERT_ID", IDENTITY,
-                0, Value.LONG);
+                VAR_ARGS, Value.LONG);
```
You need a separate function number for `LAST_INSERT_ID` because after your changes it is different from other functions mapped to `IDENTITY`.

x7FFFFFFF replied:
Why LAST_INSERT_ID are in Functions, but not in FunctionsMySQL?

katzyn replied:
Because it was implemented a long time ago much earlier than support for mode-specific functions.

Yes, it's better to remove the mapping for it from `Function` and move your implementation to `FunctionsMySQL`. We don't need this function in other modes.

You can use `SET MODE MySQL;` command in `last-insert-id.sql` to change the mode. It will be necessary, because functions from `FunctionsMySQL` are not available by default.

katzyn commented on h2/src/main/org/h2/expression/function/Function.java:
```diff
@@ -899,7 +899,13 @@ private Value getSimpleValue(Session session, Value v0, Expression[] args,
+                result = session.getLastIdentity();
+            } else {
+                final Value value = getNullOrValue(session, args, values, 0);
+                session.setLastIdentity(value);
```
Use parameter `v0` here.

Also I think that it's reasonable to convert the argument to `Value.LONG` (`v0.convertTo(Value.LONG)`) to ensure that preserved last identity will have a valid data type.

BTW, what exactly MySQL does if `NULL` is passed as an argument?

x7FFFFFFF replied:
BTW, what exactly MySQL does if NULL is passed as an argument?


MariaDB [test1]> SELECT LAST_INSERT_ID(NULL);
+----------------------+
| LAST_INSERT_ID(NULL) |
+----------------------+
|                 NULL |
+----------------------+
1 row in set (0.00 sec)

MariaDB [test1]> SELECT LAST_INSERT_ID();
+------------------+
| LAST_INSERT_ID() |
+------------------+
|                0 |
+------------------+
1 row in set (0.00 sec)
`
version    10.0.38-MariaDB

katzyn commented on h2/src/test/org/h2/test/scripts/TestScript.java:
```diff
@@ -201,7 +201,7 @@ public void test() throws Exception {
                 "memory-free", "memory-used", "nextval", "nullif", "nvl2",
                 "readonly", "rownum", "schema", "scope-identity", "session-id",
-                "set", "table", "transaction-id", "truncate-value", "unnest", "user" }) {
+                "set", "table", "transaction-id", "truncate-value", "unnest", "user", "last-insert-id" }) {
```
This list is sorted alphabetically.

katzyn commented on h2/src/test/org/h2/test/scripts/functions/system/last-insert-id.sql:
```diff
@@ -0,0 +1,51 @@
+create memory table sequence (id INT NOT NULL AUTO_INCREMENT, title varchar(255));
```
Copy a copyright header from any other script.

katzyn commented on h2/src/test/org/h2/test/scripts/functions/system/last-insert-id.sql:
```diff
@@ -0,0 +1,51 @@
+> update count: 1
+
+--UPDATE sequence SET id=LAST_INSERT_ID(id+1);
+--> update count: 1
```
If these lines aren't needed, they should be removed.

katzyn commented on h2/src/test/org/h2/test/scripts/functions/system/last-insert-id.sql:
```diff
@@ -0,0 +1,51 @@
+> L
+> -
+> 2
+> rows: 1
```
You can use short notation
```
>> 2
```
here and below.

paramonov-alex committed:
Issue #308
https://github.com/h2database/h2database/issues/308
Mode MySQL and LAST_INSERT_ID with argument
Fix

katzyn commented:
Thanks!
